File: src/opentrace_ml/gpx.py
"""Load timestamped GPS traces from GPX files."""

from __future__ import annotations

from datetime import datetime, timezone
from itertools import pairwise
from pathlib import Path
from xml.etree import ElementTree

from .models import GeoPoint


def _parse_timestamp(value: str) -> datetime:
    """Parse an ISO-8601 timestamp and normalize aware values to UTC."""

    parsed = datetime.fromisoformat(value.strip().replace("Z", "+00:00"))
    if parsed.tzinfo is not None:
        return parsed.astimezone(timezone.utc).replace(tzinfo=None)
    return parsed


def load_gpx_points(path: str | Path) -> list[GeoPoint]:
    """Load GPX track points in document order with trip-relative timestamps.

    Timezone-aware timestamps are normalized to UTC. Timezone-naive timestamps
    are interpreted consistently with one another. Track order is preserved and
    must also be chronological.
    """

    try:
        root = ElementTree.parse(path).getroot()
    except ElementTree.ParseError as error:
        raise ValueError("GPX file is not valid XML") from error

    points: list[tuple[float, float, datetime]] = []
    for element in root.iter():
        if element.tag.rsplit("}", 1)[-1] != "trkpt":
            continue
        latitude = element.get("lat")
        longitude = element.get("lon")
        timestamp = next(
            (child.text for child in element if child.tag.rsplit("}", 1)[-1] == "time"),
            None,
        )
        if latitude is None or longitude is None:
            raise ValueError("GPX track point is missing latitude or longitude")
        if not timestamp:
            raise ValueError("GPX track point is missing a timestamp")
        try:
            points.append((float(latitude), float(longitude), _parse_timestamp(timestamp)))
        except (TypeError, ValueError) as error:
            raise ValueError("GPX track point has an invalid timestamp or coordinate") from error

    if not points:
        raise ValueError("GPX file contains no track points")
    if any(current[2] > following[2] for current, following in pairwise(points)):
        raise ValueError("GPX track points must be ordered by timestamp")

    start = points[0][2]
    return [GeoPoint(lat, lon, (timestamp - start).total_seconds()) for lat, lon, timestamp in points]

File: go/gpx.go
package opentrace

import (
	"encoding/xml"
	"errors"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
	"time"
)

type gpxTrackPoint struct {
	Latitude  string `xml:"lat,attr"`
	Longitude string `xml:"lon,attr"`
	Time      string `xml:"time"`
}

// LoadGPXFile parses timestamped track points from a GPX file.
func LoadGPXFile(path string) ([]GeoPoint, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("opentrace: open GPX: %w", err)
	}
	defer file.Close()
	return LoadGPX(file)
}

// LoadGPX decodes GPX track points in document order using O(n) memory. Aware
// timestamps are normalized to UTC; naive timestamps are treated as UTC.
func LoadGPX(reader io.Reader) ([]GeoPoint, error) {
	if reader == nil {
		return nil, errors.New("opentrace: GPX reader is nil")
	}
	decoder := xml.NewDecoder(reader)
	type absolutePoint struct {
		latitude, longitude float64
		timestamp           time.Time
	}
	absolute := make([]absolutePoint, 0)
	depth, roots := 0, 0
	for {
		token, err := decoder.Token()
		if errors.Is(err, io.EOF) {
			break
		}
		if err != nil {
			return nil, fmt.Errorf("opentrace: GPX is not valid XML: %w", err)
		}
		if _, ok := token.(xml.EndElement); ok {
			depth--
			continue
		}
		if data, ok := token.(xml.CharData); ok && depth == 0 && strings.TrimSpace(string(data)) != "" {
			return nil, errors.New("opentrace: unexpected text outside GPX root")
		}
		start, ok := token.(xml.StartElement)
		if ok {
			if depth == 0 {
				roots++
				if roots != 1 || start.Name.Local != "gpx" {
					return nil, errors.New("opentrace: expected one GPX root element")
				}
			}
			depth++
		}
		if !ok || start.Name.Local != "trkpt" {
			continue
		}
		var point gpxTrackPoint
		if err := decoder.DecodeElement(&point, &start); err != nil {
			return nil, fmt.Errorf("opentrace: decode GPX track point: %w", err)
		}
		depth-- // DecodeElement consumed this element's end token.
		latitude, err := strconv.ParseFloat(strings.TrimSpace(point.Latitude), 64)
		if err != nil {
			return nil, errors.New("opentrace: GPX track point has invalid latitude")
		}
		longitude, err := strconv.ParseFloat(strings.TrimSpace(point.Longitude), 64)
		if err != nil {
			return nil, errors.New("opentrace: GPX track point has invalid longitude")
		}
		timestamp, err := parseGPXTimestamp(point.Time)
		if err != nil {
			return nil, errors.New("opentrace: GPX track point has an invalid timestamp")
		}
		absolute = append(absolute, absolutePoint{latitude, longitude, timestamp})
	}
	if len(absolute) == 0 {
		return nil, errors.New("opentrace: GPX contains no track points")
	}
	start := absolute[0].timestamp
	points := make([]GeoPoint, len(absolute))
	for i, point := range absolute {
		if i > 0 && point.timestamp.Before(absolute[i-1].timestamp) {
			return nil, errors.New("opentrace: GPX track points must be ordered by timestamp")
		}
		points[i] = GeoPoint{
			Latitude:  point.latitude,
			Longitude: point.longitude,
			TimestampSeconds: float64(point.timestamp.Unix()-start.Unix()) +
				float64(point.timestamp.Nanosecond()-start.Nanosecond())/1e9,
		}
		if err := points[i].Validate(); err != nil {
			return nil, fmt.Errorf("opentrace: GPX track point %d: %w", i, err)
		}
	}
	return points, nil
}

func parseGPXTimestamp(value string) (time.Time, error) {
	trimmed := strings.TrimSpace(value)
	if trimmed == "" {
		return time.Time{}, errors.New("missing timestamp")
	}
	if parsed, err := time.Parse(time.RFC3339Nano, trimmed); err == nil {
		return parsed.UTC(), nil
	}
	for _, layout := range []string{"2006-01-02T15:04:05.999999999", "2006-01-02T15:04:05"} {
		if parsed, err := time.ParseInLocation(layout, trimmed, time.UTC); err == nil {
			return parsed, nil
		}
	}
	return time.Time{}, errors.New("invalid timestamp")
}
